homeassistant/components/tahoma/tahoma_device.py

```diff
@@ -0,0 +1,169 @@
+
+from .const import DOMAIN
+from .coordinator import TahomaDataUpdateCoordinator
+
+ATTR_RSSI_LEVEL = "rssi_level"
+
+CORE_AVAILABILITY_STATE = "core:AvailabilityState"
+CORE_BATTERY_STATE = "core:BatteryState"
+CORE_MANUFACTURER_NAME_STATE = "core:ManufacturerNameState"
+CORE_MODEL_STATE = "core:ModelState"
+CORE_RSSI_LEVEL_STATE = "core:RSSILevelState"
+CORE_SENSOR_DEFECT_STATE = "core:SensorDefectState"
+CORE_STATUS_STATE = "core:StatusState"
+
+STATE_AVAILABLE = "available"
+STATE_BATTERY_FULL = "full"
+STATE_BATTERY_NORMAL = "normal"
+STATE_BATTERY_LOW = "low"
+STATE_BATTERY_VERY_LOW = "verylow"
+STATE_DEAD = "dead"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TahomaDevice(CoordinatorEntity, Entity):
+    """Representation of a TaHoma device entity."""
+
+    def __init__(self, device_url: str, coordinator: TahomaDataUpdateCoordinator):
+        """Initialize the device."""
+        super().__init__(coordinator)
+        self.device_url = device_url
+
+    @property
+    def device(self) -> Device:
+        """Return TaHoma device linked to this entity."""
+        return self.coordinator.data[self.device_url]
+
+    @property
+    def name(self) -> str:
+        """Return the name of the device."""
+        return self.device.label
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self.device.available
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return self.device.deviceurl
+
+    @property
+    def assumed_state(self) -> bool:
+        """Return True if unable to access real state of the entity."""
+        return self.device.states is None or len(self.device.states) == 0
+
+    @property
+    def device_state_attributes(self) -> Dict[str, Any]:
+        """Return the state attributes of the device."""
+        attr = {
+            "ui_class": self.device.ui_class,
+            "widget": self.device.widget,
+            "controllable_name": self.device.controllable_name,
+        }
+
+        if self.has_state(CORE_RSSI_LEVEL_STATE):
+            attr[ATTR_RSSI_LEVEL] = self.select_state(CORE_RSSI_LEVEL_STATE)
+
+        if self.has_state(CORE_BATTERY_STATE):
+            battery_state = self.select_state(CORE_BATTERY_STATE)
+
+            if battery_state == STATE_BATTERY_FULL:
+                battery_state = 100
+            elif battery_state == STATE_BATTERY_NORMAL:
+                battery_state = 75
+            elif battery_state == STATE_BATTERY_LOW:
+                battery_state = 25
+            elif battery_state == STATE_BATTERY_VERY_LOW:
+                battery_state = 10
+
+            attr[ATTR_BATTERY_LEVEL] = battery_state
+
+        if self.select_state(CORE_SENSOR_DEFECT_STATE) == STATE_DEAD:
+            attr[ATTR_BATTERY_LEVEL] = 0
+
+        if self.device.attributes:
+            for attribute in self.device.attributes:
+                attr[attribute.name] = attribute.value
```

Comment:
Please set the attribute keys explicitly so we can see what they are and their formatting. Eg by creating a constant dict that maps state attribute key to device attribute key, and iterating over that dict while looking up the device attribute values.

Reply:
All attributes removed for now, others could be added in a follow up PR.